### PR TITLE
[stable] Add IPFS Chart

### PR DESCRIPTION
**Pull request by yuvipanda**
Adds a chart that lets you install [IPFS](https://ipfs.io) with a StatefulSet.

### CODE REVIEW

File: stable/ipfs/templates/statefulset.yaml
Comment (victorb) [inline stable/ipfs/templates/statefulset.yaml]: Probably want to use `ipfs/go-ipfs` instead of from Juan's account, and also lock it to a version, `v0.4.9` is latest available in Docker Hub.

Comment (yuvipanda) [reply stable/ipfs/templates/statefulset.yaml]: Good catch! Fixed it :)

File: stable/ipfs/templates/service.yaml
Comment (unguiculus) [inline stable/ipfs/templates/service.yaml]: Why would you want to do that? This should just be `{{ template "fullname" . }}`.

File: stable/ipfs/templates/statefulset.yaml
Comment (unguiculus) [inline stable/ipfs/templates/statefulset.yaml]: Name should be `{{ template "fullname" . }}`.

File: stable/ipfs/templates/statefulset.yaml
Comment (unguiculus) [inline stable/ipfs/templates/statefulset.yaml]: Use this `{{ template "name" . }}` for the `app` label.

File: stable/ipfs/templates/statefulset.yaml
Comment (unguiculus) [inline stable/ipfs/templates/statefulset.yaml]: Can you explain the reason behind this?

File: stable/ipfs/templates/statefulset.yaml
Comment (unguiculus) [inline stable/ipfs/templates/statefulset.yaml]: Make image configurable.

File: stable/ipfs/values.yaml
Comment (unguiculus) [inline stable/ipfs/values.yaml]: Don't specify defaults for resources.

File: stable/ipfs/templates/service.yaml
Comment (unguiculus) [inline stable/ipfs/templates/service.yaml]: Use standard labels everywhere. See:
https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/labels.md

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/ipfs/templates/statefulset.yaml
Comment (unguiculus) [inline stable/ipfs/templates/statefulset.yaml]: Could you make this more flexible and make persistence optional?
See e. g.: https://github.com/kubernetes/charts/blob/master/stable/mongodb-replicaset/templates/mongodb-statefulset.yaml#L175-L199

File: stable/ipfs/templates/service.yaml
Comment (yuvipanda) [reply stable/ipfs/templates/service.yaml]: Because I want to be allow users to override service names. In this case, end users will be asked to hit the DNS entry for the service name, and that should be overrideable so it can be something like 'ipfs' so other users in the same namespace can just specify 'ipfs'. The default is still the behavior seen in the current charts in the repo.

File: stable/ipfs/templates/service.yaml
Comment (yuvipanda) [reply stable/ipfs/templates/service.yaml]: I just got this from helm create, will read and update!

File: stable/ipfs/templates/statefulset.yaml
Comment (yuvipanda) [reply stable/ipfs/templates/statefulset.yaml]: Yes - the image needs to write to the PVC to function, and it runs as UID 1000 (rather than root).

File: stable/ipfs/templates/service.yaml
Comment (yuvipanda) [reply stable/ipfs/templates/service.yaml]: Done!

File: stable/ipfs/templates/statefulset.yaml
Comment (yuvipanda) [reply stable/ipfs/templates/statefulset.yaml]: Done!

File: stable/ipfs/templates/statefulset.yaml
Comment (yuvipanda) [reply stable/ipfs/templates/statefulset.yaml]: Done!

File: stable/ipfs/templates/statefulset.yaml
Comment (yuvipanda) [reply stable/ipfs/templates/statefulset.yaml]: Done

File: stable/ipfs/values.yaml
Comment (yuvipanda) [reply stable/ipfs/values.yaml]: Done

File: stable/ipfs/templates/statefulset.yaml
Comment (yuvipanda) [reply stable/ipfs/templates/statefulset.yaml]: I could if it is a requirement for merging :) but it's not particularly useful without persistent volumes however - but I guess so is the case for MongoDB!

Comment (yuvipanda) [reply stable/ipfs/templates/statefulset.yaml]: Done! Optional but enabled by default!

File: stable/ipfs/values.yaml
Comment (unguiculus) [inline stable/ipfs/values.yaml]: Remove `nameOverride: null`

File: stable/ipfs/templates/service.yaml
Comment (unguiculus) [inline stable/ipfs/templates/service.yaml:21]: Requires an additional selector: `release: {{ .Release.Name }}`

File: stable/ipfs/templates/statefulset.yaml
Comment (unguiculus) [inline stable/ipfs/templates/statefulset.yaml:16]: At least the `release` label should be added here as well.

File: stable/ipfs/templates/statefulset.yaml
Comment (unguiculus) [inline stable/ipfs/templates/statefulset.yaml]: `if` not necessary. Add default to `values.yaml`:

`resources: {}`

File: stable/ipfs/values.yaml
Comment (unguiculus) [inline stable/ipfs/values.yaml]: Remove `nameOverride`.

File: stable/ipfs/templates/statefulset.yaml
Comment (unguiculus) [inline stable/ipfs/templates/statefulset.yaml]: I don't think this is necessary. Will test this. Otherwise we'd have this problem with lots of charts. I think this is only a problem if you mount a volume from the host, which is not the case.

File: stable/ipfs/templates/_helpers.tpl
Comment (unguiculus) [inline stable/ipfs/templates/_helpers.tpl:26]: Add truncation.

File: stable/ipfs/README.md
Comment (unguiculus) [inline stable/ipfs/README.md]: Add missing config entries.

File: stable/ipfs/templates/service.yaml
Comment (unguiculus) [inline stable/ipfs/templates/service.yaml:11]: Statefulsets require a headless service. This means you should always have this:

```yaml
type: ClusterIP
clusterIP: None
```

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/ipfs/templates/service.yaml
Comment (yuvipanda) [reply stable/ipfs/templates/service.yaml:11]: Does the service need to be headless? We also need a service anyway, and this paradigm works well - ipfs.namespace round robins to one of them, and you can do ipfs-0.ipfs or ipfs-1.ipfs for each one if you need.

File: stable/ipfs/templates/statefulset.yaml
Comment (yuvipanda) [reply stable/ipfs/templates/statefulset.yaml]: I ran into permissionDenied errors when I tried (on GKE). Perhaps the runAsUser isn't necessary but fs is?

File: stable/ipfs/templates/service.yaml
Comment (yuvipanda) [reply stable/ipfs/templates/service.yaml:21]: Done

File: stable/ipfs/values.yaml
Comment (yuvipanda) [reply stable/ipfs/values.yaml]: Done

File: stable/ipfs/values.yaml
Comment (yuvipanda) [reply stable/ipfs/values.yaml]: Done

File: stable/ipfs/templates/statefulset.yaml
Comment (yuvipanda) [reply stable/ipfs/templates/statefulset.yaml]: Done

File: stable/ipfs/templates/statefulset.yaml
Comment (yuvipanda) [reply stable/ipfs/templates/statefulset.yaml:16]: Done

**Comment by unguiculus** [comment]
@k8s-bot ok to test

**Review by unguiculus — APPROVED** [review]

File: stable/ipfs/templates/_helpers.tpl
Comment (yuvipanda) [reply stable/ipfs/templates/_helpers.tpl:26]: On second thought, am not sure this needs truncation. The user is responsible for setting it, and the default fullname handles truncation already.

File: stable/ipfs/README.md
Comment (yuvipanda) [reply stable/ipfs/README.md]: Done!

**Comment by yuvipanda** [comment]
Anything else I need to do before I can get it merged? :)

File: stable/ipfs/templates/NOTES.txt
Comment (unguiculus) [inline stable/ipfs/templates/NOTES.txt]: `service.externalPort` is not used anywhere. Thus, a default installation prints this:

```
If you want to connect to it from your local computer, you can find a URL to connect with the
following (for the gateway service):
  export POD_NAME=$(kubectl get pods --namespace default -l "app=wiggly-alpaca-ipfs" -o jsonpath="{.items[0].metadata.name}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl port-forward $POD_NAME 8080:
```

Also note that the label selector is wrong. Use this: `-l "app={{ template "name" . }},release={{ .Release.Name }}"`

**Review by unguiculus — CHANGES_REQUESTED** [review]
I deployed it on GKE. Everything seems to come up successfully. The log looks ok. When I access it on port 8080 with port-forwarding, I get a 404.

**Comment by unguiculus** [comment]
Marking as stale. Please update within one week.

**Comment by yuvipanda** [comment]
Whoops, shall do!

On Jul 2, 2017 11:06 AM, "Reinhard Nägele" <notifications@github.com> wrote:

Marking as stale. Please update within one week.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/charts/pull/1192#issuecomment-312507375>,
or mute
the thread

**Comment by yuvipanda** [comment]
The 404 is expected, since it's only an API Gateway service. If you access, for example, /ipfs/Qmb8wsGZNXt5VXZh1pEmYynjB6Euqpq3HYyeAdw2vScTkQ, it'll lead you to the same page as https://ipfs.io/ipfs/Qmb8wsGZNXt5VXZh1pEmYynjB6Euqpq3HYyeAdw2vScTkQ. It can also serve a particular hash based on a Host Header. For example, if you do a `curl -H 'Host: ipfs.io' <ip>` you'll actually get back the contents of ipfs.io landing page.

File: stable/ipfs/templates/NOTES.txt
Comment (yuvipanda) [reply stable/ipfs/templates/NOTES.txt]: Updated!

**Comment by unguiculus** [comment]
@yuvipanda 

Two things about the `Chart.yaml` I noticed after merging:
- The maintainer name should be a Github user
- The `appVersion` field should be added

Would you mind creating another PR for this?

**Comment by yuvipanda** [comment]
@unguiculus I sent #1469. Thanks!

**Pull request merged**
